File: packages/imanikurd/src/library.ts
import { loadData, loadDataAsync } from "./loadData.js";

export interface LibraryItem {
  id: number;
  title: string;
  author?: string;
  url?: string;
  description?: string;
  [key: string]: unknown;
}

interface LibraryData {
  books: LibraryItem[];
}

let libraryCache: LibraryItem[] | null = null;

function getLibraryData(): LibraryItem[] {
  if (!libraryCache) {
    const raw = loadData<LibraryData>("library.json");
    libraryCache = raw.books || [];
  }
  return libraryCache;
}

async function getLibraryDataAsync(): Promise<LibraryItem[]> {
  if (!libraryCache) {
    const raw = await loadDataAsync<LibraryData>("library.json");
    libraryCache = raw.books || [];
  }
  return libraryCache;
}

/** Get Islamic library book list */
export function getLibrary(): LibraryItem[] {
  return getLibraryData();
}

/** Get Islamic library book list asynchronously */
export async function getLibraryAsync(): Promise<LibraryItem[]> {
  return getLibraryDataAsync();
}

/** Get a library item by id */
export function getLibraryItem(id: number): LibraryItem | undefined {
  return getLibraryData().find((item) => item.id === id);
}

/** Get a library item by id asynchronously */
export async function getLibraryItemAsync(id: number): Promise<LibraryItem | undefined> {
  const data = await getLibraryDataAsync();
  return data.find((item) => item.id === id);
}

/** Search library by title or author */
export function searchLibrary(query: string): LibraryItem[] {
  const normalized = query.trim().toLowerCase();
  if (!normalized) return [];

  return getLibraryData().filter(
    (item) =>
      String(item.title || "").toLowerCase().includes(normalized) ||
      String(item.author || "").toLowerCase().includes(normalized)
  );
}

/** Search library by title or author asynchronously */
export async function searchLibraryAsync(query: string): Promise<LibraryItem[]> {
  const normalized = query.trim().toLowerCase();
  if (!normalized) return [];

  const data = await getLibraryDataAsync();
  return data.filter(
    (item) =>
      String(item.title || "").toLowerCase().includes(normalized) ||
      String(item.author || "").toLowerCase().includes(normalized)
  );
}

File: packages/imanikurd_flutter/lib/src/library.dart
import 'load_data.dart';
import 'models.dart';

List<LibraryItem>? _libraryCache;

Future<List<LibraryItem>> _getLibraryData() async {
  if (_libraryCache != null) return _libraryCache!;
  final raw = await loadDataAsync<Map<String, dynamic>>('library.json');
  final books = raw['books'] as List<dynamic>? ?? [];
  _libraryCache = books.map((e) => LibraryItem.fromJson(e as Map<String, dynamic>)).toList();
  return _libraryCache!;
}

Future<List<LibraryItem>> getLibrary() => _getLibraryData();

Future<LibraryItem?> getLibraryItem(int id) async {
  final data = await _getLibraryData();
  for (final item in data) {
    if (item.id == id) return item;
  }
  return null;
}

Future<List<LibraryItem>> searchLibrary(String query) async {
  final normalized = query.trim().toLowerCase();
  if (normalized.isEmpty) return [];
  final data = await _getLibraryData();
  return data
      .where(
        (item) =>
            item.title.toLowerCase().contains(normalized) ||
            (item.author ?? '').toLowerCase().contains(normalized),
      )
      .toList();
}
